fix linker-built array padding issues on 64-bit targets

npitre opened the pull request:
Please see individual commits for the details. The second one depends
on the first so I'm submitting them together.

npitre committed:
DEVICE_DEFINE(): properly align struct device instances

The DEVICE_DEFINE() macro creates instances of struct device that are
gathered into a contiguous list by the linker. However, some assemblers
pad sections to the next 16-byte boundary or so by default, screwing up
the list walk in z_sys_device_do_config_level(). This is especially
true for 64-bit compilation where sizeof(struct device) isn't a
multiple of 16.

Enforcing an alignment at the linker level would solve this issue when
instances of struct device are gathered from different object files.
However it doesn't solve it when multiple instances are created within
the same object file where the first instance still has a gap with the
next instance, as the assembler does add padding upon section switch
even though the object file ends up with a single section with both
instances. In that case the linker would get rid of the trailing padding
only, leaving the inner gaps between instances in place.

The actual fix is to provide an explicit alignment attribute to the
section for every instances, using __alignof(struct device) which is
the alignment expected by the compiler for that structure.

This also means that the x86_64 workaround in the struct device
definition may go as the "edge case" it refers to is now properly
handled.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
linker generated list: explicit alignment on data definitions

The alignment fix on struct device definitions should be done to all
such linker list tricks. Let's abstract the declaration plus alignment
with a macro and apply it to all concerned cases.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

andyross reviewed (commented):
Seems like the __decl_align trick used here might still be a little fragile (not that the earlier code wasn't, but...).   This sounds like a spot where we really want to stop and fix the design.

Really we have a situation here where we're implementing a specific pattern in a zillion places: have the linker pack a contiguous array of a speciifc struct type for us with start/end boundaries over which we can iterate at runtime.  And we have a particular failure mode: the linker stride for those objects may not match the runtime size of the object we get from sizeof().

Solution Number 0: Ignore the issue.  Everything gets packed in 32 bit units anyway, right?  Works everywhere, except when it doesn't.

Solution Number 1: (as of the x86_64 port) Mostly ignore the issue, and go in and hack things with padding for situations where the strides differ.  Works everywhere, once we introduce diagnose and fix the alignment bugs.

Solution Number 2: (this patch) Don't ignore the issue.  Go in everywhere we use this trick and declaratively align the structs to their link time alignments.  Works everywhere.  Requires a particular idiom and the knowledge of how __decl_align() differs from aligned(__alignof(...).

How about coming up with a particular tag, e.g. Z_LINK_TIME_ARRAY or something, that we can apply to a struct declaration to get the correct alignment?  And maybe some code somewhere (maybe in the linker, or possibly a python script) that can validate that the strides in the resulting image match what the code was compiled to expect.  And while we're at it, a Z_LINK_TIME_ARRAY_FOREACH iterator to replace the existing hand-cooked stuff would be really nice.

It's not like I think this is the wrong fix, just that we have a really messy pattern in the code and if we're going to dive in and pervasively change the pattern, let's take the time to pick the right pattern.

npitre commented:
@andyross commented:

> How about coming up with a particular tag, e.g. Z_LINK_TIME_ARRAY or 
> something, that we can apply to a struct declaration to get the 
> correct alignment?

You cannot just stamp a standalone tag on a structure alongside its 
declaration. You also need its type to provide the appropriate 
alignment. This __decl_align() is the most generic method I could come 
up with.

> And maybe some code somewhere (maybe in the linker, or possibly a 
> python script) that can validate that the strides in the resulting 
> image match what the code was compiled to expect.

The linker cannot infer the needed alignment. That's the compiler's job 
to determine and provide it to the linker. Here we're outsmarting the 
compiler somewhat which is why this issue came up.

And deferring responsibility to an ad hoc python script or the like 
feels wrong to me. If you're worried the probability for the interface 
to be misused is high enough to justify such a safety net, then the 
interface is probably wrong in the first place.

I think that a better solution is to simply merge the __in_section() and 
__decl_align() together so the later won't go without the former. I 
almost went that way but the number of changed lines would have been 
larger and I didn't want to commit to that without some feedback.

That would have been something like:

  __declare_in_section(type, name, section, subsection, subsubsect)

Let me know.

aescolar reviewed (commented):
@ceolin Doesn't MISRA-C discourage macros starting with `__` ?

npitre committed:
linker generated list: introduce Z_STRUCT_SECTION_ITERABLE()

This convenience macro wraps Z_DECL_ALIGN() and __in_section() to
simplify static definitions of structure instances gathered in dedicated
sections. Most of the time those go together, and the section name is
already closely related to the struct type, so abstracting things behind
a simpler interface reduces probability of mistakes and makes the code
clearer. A few input section names have been adjusted accordingly.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
linker generated list: provide an iterator to simplify list access

Given that the section name and boundary simbols can be inferred from
the struct object name, it makes sense to create an iterator that
abstracts away the access details and reduce the possibility for
mistakes.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
On Wed, 29 May 2019, Andy Ross wrote:

> Seems like the __decl_align trick used here might still be a little 
> fragile (not that the earlier code wasn't, but...).  This sounds like 
> a spot where we really want to stop and fix the design.

So I updated this PR with the solution I was thinking of to make the whole scheme less fragile.

I also added the iterator you alluded to and applied to all the easy cases.

andyross commented on include/toolchain/common.h:
```diff
@@ -144,4 +144,48 @@
+ * between sections within the same file even if it merges many such segments
+ * into a single section in the end.
+ */
+#define __decl_align(type) __aligned(__alignof(type)) type
```
As mentioned by someone else, MISRA forbids us from introducing new symbols with a prepended underscore.  Should be "Z_DECL_ALIGN" by current convention.

andyross commented on include/toolchain/common.h:
```diff
@@ -144,4 +144,48 @@
+ * _<struct_type>_list_end symbol to mark the start and the end of the
+ * list of struct objects to iterate over.
+ */
+#define iterate_struct_from_section(struct_type, iterator) \
```
Similarly, common Zephyrish would name this "Z_STRUCT_SECTION_FOREACH" (or per my suggestion: Z_ITERABLE_STATIC_FOREACH).

andyross reviewed (approved):
Yeah, this looks great and is exactly what I imagined we should be able to have the API look like.  Some nitpicks on naming that maybe someone else wants to start a fight about, but I'd be very happy to see this merge as-is.

andyross commented on include/toolchain/common.h:
```diff
@@ -144,4 +144,48 @@
+ * The section name is the struct type prepended with an underscore.
+ * The subsection is "static" and the subsubsection is the variable name.
+ */
+#define declare_struct_in_section(struct_type, name) \
```
(whoops, dropped this comment, sorry)

This I like very much!

Style: Generally, zephyr code will put the verb "declare" at the end
of the symbol name, use upper case for macros, and add a z_ or Z_ prefix to
all internal symbols.  So maybe Z_STRUCT_SECTION_DECLARE() might be more conventional.

Or if I can suggest: maybe calling the feature "ITERABLE_STATIC" better captures the magic at work.  The point isn't that we're using structs (this trick works for any fixed-size data type) or that they're going in a particular section, it's that the link magically makes them into a list for us at runtime no matter where they might be.

npitre commented:
On Mon, 3 Jun 2019, Andy Ross wrote:

> This I like very much!

Thanks.

> Or if I can suggest: maybe calling the feature "ITERABLE_STATIC" better captures the magic at work.

Given that the type is assumed to be a struct by the macro, I think the 
name should still carry that. Otherwise I've applied your suggestions.

tyrantlizard77 reviewed (approved):
love it!
